Prototype version 2 to support NCCL zero copy

minsii opened the pull request:
Manually port from D50726970 stack for OSS testing, not for review

Skylion007 commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -969,6 +975,11 @@ class DeviceCachingAllocator {
 
+  void attachAllocatorTraceTracker(AllocatorTraceTracker tracker) {
+    enable_trace_tracker_ = true;
+    trace_trackers_.emplace_back(tracker);
```
```suggestion
    trace_trackers_.emplace_back(std::move(tracker));
```

minsii commented:
The cache allocator commit is reviewed via https://github.com/pytorch/pytorch/pull/112238

The intention of this PR is to test NCCL PG level change together with the cache allocator commit.

minsii committed:
[c10d] use allocator trace callbacks for NCCL PG register

Summary:
We need to register all cache segments allocated by allocator, so that NCCL can apply zero copy algorithms at collective and point-to-point operations.

How to track and register all cache segments:
- It registers a register and a deregister hook to cache allocator as action tracker callbacks, tracking SEGMENT_ALLOC and SEGMENT_FREE trace entries, respectively. When SEGMENT_ALLOC is tracked, the register hook will register to the PG's communicators on the same device. Similarly, when SEGMENT_FREE is tracked, the deregister hook handles deregistration before cudaFree.
- When a new NCCL communicator is created, it dumps the snapspot from cache allocator to register all existing cache segments at once.
- When a NCCL communicator is aborted, it deregisters all segments that have been registered by this communicator

Test Plan: See test in D50726971

Reviewed By: wconstab

Differential Revision: D50726970

minsii commented:
Close this prototype PR as the final version https://github.com/pytorch/pytorch/pull/112850 has exported
